### PR TITLE
UCT/UD: Add NAK response

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by brminich** [comment]
TMO can be relevant
```
[2020-06-18T08:48:57.058Z] Running org.openucx.jucx.UcpEndpointTest

[2020-06-18T08:54:03.684Z] [1592470439.126092] [hpc-test-node-upstream-01:16738:0] jucx_common_def.cc:191  UCX  ERROR JUCX: request error: Endpoint timeout

[2020-06-18T08:54:03.684Z] [1592470439.126624] [hpc-test-node-upstream-01:16738:0] jucx_common_def.cc:191  UCX  ERROR JUCX: request error: Endpoint timeout

[2020-06-18T08:59:10.530Z] [1592470742.074281] [hpc-test-node-upstream-01:16738:0] jucx_common_def.cc:191  UCX  ERROR JUCX: request error: Endpoint timeout

[2020-06-18T08:59:20.472Z] [1592470759.515703] [hpc-test-node-upstream-01:16738:0] jucx_common_def.cc:191  UCX  ERROR JUCX: request error: Endpoint timeout

[2020-06-18T08:59:20.472Z] [1592470759.516674] [hpc-test-node-upstream-01:16738:0]    ib_mlx5_log.c:143  UCX  ERROR Transport retry count exceeded on mlx5_bond_0:1/RoCE (synd 0x15 vend 0x81 hw_synd 0/0)

[2020-06-18T08:59:20.472Z] [1592470759.516674] [hpc-test-node-upstream-01:16738:0]    ib_mlx5_log.c:143  UCX  ERROR RC QP 0xd0cb wqe[2]: SEND --e [va 0x7f62bce0ff00 len 4106 lkey 0x1af09]
```

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -57,6 +57,12 @@ static void uct_ud_ep_resend_start(uct_ud_iface_t *iface, uct_ud_ep_t *ep)
     uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_RESEND);
 }
 
+static void uct_ud_ep_resend_end(uct_ud_ep_t *ep)
+{
+    uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_RESEND);
+    ep->tx.naked = 0;
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:63]: UCT_UD_EP_FLAG_TX_NACKED

File: src/uct/ib/ud/base/ud_ep.c [hunk 2]
```diff
@@ -316,7 +322,7 @@ static void uct_ud_ep_timer(ucs_wtimer_t *self)
     if (diff > 3 * iface->tx.tick) {
         ucs_trace("scheduling resend now: %lu last_send: %lu diff: %lu tick: %lu",
                   now, last_send, diff, ep->tx.tick);
-        uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_ACK_REQ);
+        uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_ACK_REQ);
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:325]: 1. unite if/else
2. call ca_drop also when got NAK

File: src/uct/ib/ud/base/ud_ep.c [hunk 3]
```diff
@@ -828,6 +834,16 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
                        ep->rx.ooo_pkts.head_sn, neth->psn);
     }
 
+    if (ucs_unlikely(neth->packet_type & UCT_UD_PACKET_FLAG_NAK)) {
+        ep->tx.naked = 1;
+        goto out;
+    }
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:840]: can we add under if in line 847 for opt? are nak packets always not active messages?

File: src/uct/ib/ud/base/ud_ep.c [hunk 4]
```diff
@@ -1094,7 +1109,13 @@ static void uct_ud_ep_resend(uct_ud_ep_t *ep)
     if (UCT_UD_PSN_COMPARE(sent_skb->neth->psn, >=, ep->tx.max_psn)) {
         ucs_debug("ep(%p): out of window(psn=%d/max_psn=%d) - can not resend more",
                   ep, sent_skb ? sent_skb->neth->psn : -1, ep->tx.max_psn);
-        uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_RESEND);
+        uct_ud_ep_resend_end(ep);
+        return;
+    }
+
+    if (!ep->tx.naked && (sent_skb->len > sizeof(uct_ud_neth_t) +
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:1116]: add comment

**Comment by Artemy-Mellanox** [comment]
bot:pipe:retest

File: src/uct/ib/ud/base/ud_ep.c [hunk 5]
```diff
@@ -322,10 +322,11 @@ static void uct_ud_ep_timer(ucs_wtimer_t *self)
     if (diff > 3 * iface->tx.tick) {
         ucs_trace("scheduling resend now: %lu last_send: %lu diff: %lu tick: %lu",
                   now, last_send, diff, ep->tx.tick);
-        uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_ACK_REQ);
         uct_ud_ep_ca_drop(ep);
         uct_ud_ep_resend_start(iface, ep);
-    } else if ((diff > iface->tx.tick) && uct_ud_ep_is_connected(ep)) {
+    }
+
+    if ((diff > iface->tx.tick) && uct_ud_ep_is_connected(ep)) {
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:329]: maybe we can place the if in line 322 under this 'if'?
also the comment below: ` /* Try to request ACK or NAK twice before going into full resend mode */`

File: src/uct/ib/ud/base/ud_ep.c [hunk 6]
```diff
@@ -828,7 +835,16 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
                        ep->rx.ooo_pkts.head_sn, neth->psn);
     }
 
+    if (UCT_UD_PSN_COMPARE(neth->psn, >, ep->rx.ooo_pkts.head_sn + 1)) {
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:838]: ucs_unlikely

File: src/uct/ib/ud/base/ud_ep.c [hunk 7]
```diff
@@ -834,17 +835,16 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
...
-    }
-
-
     if (UCT_UD_PSN_COMPARE(neth->psn, >, ep->rx.ooo_pkts.head_sn + 1)) {
-        uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_NAK);
+        uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_NACK);
     }
 
     if (ucs_unlikely(!is_am)) {
+        if (ucs_unlikely(neth->packet_type & UCT_UD_PACKET_FLAG_NAK)) {
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:843]: minor: here, no need for ucs_unlikely, since line 842 already has it

File: src/uct/ib/ud/base/ud_ep.c [hunk 8]
```diff
@@ -1113,8 +1113,11 @@ static void uct_ud_ep_resend(uct_ud_ep_t *ep)
         return;
     }
 
-    if (!ep->tx.naked && (sent_skb->len > sizeof(uct_ud_neth_t) +
-        sizeof(uct_ud_ctl_hdr_t) + iface->super.addr_size)) {
+    /* stop resend if packet is larger than CREQ and there wasn't NACK from
+     * other side */
+    if (!(ep->flags & UCT_UD_EP_FLAG_TX_NACKED) && (sent_skb->len >
+        sizeof(uct_ud_neth_t) + sizeof(uct_ud_ctl_hdr_t) +
+        iface->super.addr_size)) {
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:1120]:
```
max_len_without_nack = sizeof(uct_ud_neth_t) + sizeof(uct_ud_ctl_hdr_t) +
                       iface->super.addr_size;
if (!(ep->flags & UCT_UD_EP_FLAG_TX_NACKED) &&
    (sent_skb->len > max_len_without_nack)) { 
 ...
```

File: test/gtest/uct/test_p2p_am.cc [hunk 9]
```diff
@@ -564,7 +564,7 @@ UCS_TEST_SKIP_COND_P(uct_p2p_am_misc, no_rx_buffs,
                                   UCT_IFACE_FLAG_CB_SYNC)))
 {
     mapped_buffer sendbuf(ucs_min(sender().iface_attr().cap.am.max_short,
-                                  10 * sizeof(uint64_t)),
+                                  5 * sizeof(uint64_t)),
```
Comment (yosefe) [inline test/gtest/uct/test_p2p_am.cc:567]: why changed?

File: src/uct/ib/ud/base/ud_ep.c [hunk 10]
```diff
@@ -1082,6 +1084,9 @@ static void uct_ud_ep_send_creq_crep(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
 static void uct_ud_ep_resend(uct_ud_ep_t *ep)
 {
     uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
+    size_t max_len_without_nack = sizeof(uct_ud_neth_t) +
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:1087]: align = with prev line

File: test/gtest/uct/test_p2p_am.cc [hunk 11]
```diff
@@ -595,7 +595,7 @@ UCS_TEST_SKIP_COND_P(uct_p2p_am_misc, no_rx_buffs,
     ucs_log_pop_handler();
 
     /* check that now the sender is able to send */
-    EXPECT_EQ(UCS_OK, send_with_timeout(sender_ep(), sendbuf, recvbuf, 6));
+    EXPECT_EQ(UCS_OK, send_with_timeout(sender_ep(), sendbuf, recvbuf, 10));
```
Comment (yosefe) [inline test/gtest/uct/test_p2p_am.cc:598]: why changed 6 -> 10 ?

File: src/uct/ib/ud/base/ud_ep.c [hunk 12]
```diff
@@ -319,16 +319,18 @@ static void uct_ud_ep_timer(ucs_wtimer_t *self)
...
+                      now, last_send, diff, ep->tx.tick);
+            uct_ud_ep_ca_drop(ep);
+            uct_ud_ep_resend_start(iface, ep);
+        }
 
-    if ((diff > iface->tx.tick) && uct_ud_ep_is_connected(ep)) {
-        /* Try to request ACK twice before going into full resend mode */
-        uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_ACK_REQ);
+        if (uct_ud_ep_is_connected(ep)) {
+            /* Try to request ACK twice before going into full resend mode */
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c:331]: can you pls add NAK:
`/* Try to request ACK or NAK twice before going into full resend mode */`

File: test/gtest/uct/test_p2p_am.cc (hunk 11 above)
Comment (Artemy-Mellanox) [reply test/gtest/uct/test_p2p_am.cc:598]: Depends on TICK

Comment (yosefe) [reply test/gtest/uct/test_p2p_am.cc:598]: maybe add comment?

**Comment by yosefe** [comment]
failure seems relevant:
http://hpc-master.lab.mtl.com:8080/blue/rest/organizations/jenkins/pipelines/ucx/runs/5865/nodes/130/steps/1855/log/?start=0
```
[2020-07-04T19:26:50.550Z] [ RUN      ] ud_verbs/uct_p2p_am_misc.no_rx_buffs/0 <ud_verbs/mlx5_bond_0:1>
[2020-07-04T19:26:50.550Z] [     INFO ] Testing component: ib
[2020-07-04T19:27:05.449Z] /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/uct/test_p2p_am.cc:599: Failure
[2020-07-04T19:27:05.449Z] Value of: send_with_timeout(sender_ep(), sendbuf, recvbuf, 10)
[2020-07-04T19:27:05.449Z]   Actual: -2
[2020-07-04T19:27:05.449Z] Expected: UCS_OK
[2020-07-04T19:27:05.449Z] Which is: 0
[2020-07-04T19:27:05.449Z] [  FAILED  ] ud_verbs/uct_p2p_am_misc.no_rx_buffs/0, where GetParam() = ud_verbs/mlx5_bond_0:1 (14554 ms)
[2020-07-04T19:27:05.449Z] [----------] 1 test from ud_verbs/uct_p2p_am_misc (14554 ms total)
[2020-07-04T19:27:05.449Z] 
```

**Comment by yosefe** [comment]
bot:pipe:retest

**Pull request merged**